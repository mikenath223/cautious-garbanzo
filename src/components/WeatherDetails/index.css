.weather-details {
  display: flex;
  gap: 1rem;
  flex-wrap: wrap;
  flex: 1 0 0;
  align-items: center;
}

.location-name {
  font-size: 1.8rem;
  line-height: 2.2rem;
}

.details-content {
  display: flex;
  flex-wrap: wrap;
  gap: 1.2rem;
  margin-top: 2.5rem;
  flex-direction: row;
}

.details-item.card {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  max-width: unset;
  min-width: 225px;
}

.font-20 {
  font-size: 20px;
  line-height: 1rem;
}

.font-16 {
  font-size: 16px;
  line-height: 0.7rem;
}

.favorites-btn {
  display: flex;
  gap: 0.5rem;
  height: 38px;
  align-items: center;
}

.flex-center {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.country-content {
  margin-top: 0.5rem;
}
