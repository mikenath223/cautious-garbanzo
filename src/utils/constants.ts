import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
