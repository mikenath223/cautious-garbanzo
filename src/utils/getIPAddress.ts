const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
