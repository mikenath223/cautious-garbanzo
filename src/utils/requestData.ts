import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}