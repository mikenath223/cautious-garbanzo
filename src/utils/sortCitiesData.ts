import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
