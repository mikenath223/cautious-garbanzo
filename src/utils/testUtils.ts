import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}