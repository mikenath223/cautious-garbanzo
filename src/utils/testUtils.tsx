import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
