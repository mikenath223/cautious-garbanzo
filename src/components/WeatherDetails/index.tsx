import { IWeatherInfo } from "../../utils/types";
import Button from "../Button";
import "./index.css";

interface IWeatherDetailsProps {
  city: IWeatherInfo;
  isFavorites: boolean;
  onAddFavorites: () => void;
  onRemoveFavorites: () => void;
}

export default function WeatherDetails({
  city,
  isFavorites = false,
  onAddFavorites,
  onRemoveFavorites,
}: IWeatherDetailsProps) {
  return (
    <section>
      <div className="weather-details">
        <img
          src={city.current.weather_icons?.[0]}
          alt=""
          className="icon-img"
        />
        <h1 className="location-name">{city.location.name}</h1>
        <Button
          className="favorites-btn"
          onClick={() =>
            !isFavorites ? onAddFavorites() : onRemoveFavorites()
          }
        >
          {!isFavorites ? (
            <span>Add to favorites</span>
          ) : (
            <span>Remove from favorites</span>
          )}
          <img
            src={
              !isFavorites
                ? "/assets/icons/unliked.png"
                : "/assets/icons/liked.png"
            }
            alt=""
            className="icon-img"
          />
        </Button>
      </div>
      <div className="country-content flex-center">
        <img src={"/assets/icons/country.png"} alt="" className="icon-img" />
        <span>{city.location.country}</span>
      </div>
      <div className="flex-center">
        <img src={"/assets/icons/state.png"} alt="" className="icon-img" />
        <span>{city.location.region}</span>
      </div>
      <ul className="details-content">
        <li className="details-item card box-shadow">
          <span className="font-20">Observation time</span>
          <span className="font-16">{city.current.observation_time}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Temperature</span>
          <span className="font-16">
            {city.current.temperature}
            <sup>o</sup>C
          </span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Weather description</span>
          <span className="font-16">
            {city.current.weather_descriptions?.[0]}
          </span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Wind speed</span>
          <span className="font-16">{city.current.wind_speed}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Wind degree</span>
          <span className="font-16">{city.current.wind_degree}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Wind direction</span>
          <span className="font-16">{city.current.wind_dir}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Pressure</span>
          <span className="font-16">{city.current.pressure}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Precipitation</span>
          <span className="font-16">{city.current.precip}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Humidity</span>
          <span className="font-16">{city.current.humidity}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Cloud cover</span>
          <span className="font-16">{city.current.cloudcover}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Feels like</span>
          <span className="font-16">{city.current.feelslike}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">UV index</span>
          <span className="font-16">{city.current.uv_index}</span>
        </li>
        <li className="details-item card box-shadow">
          <span className="font-20">Visibility</span>
          <span className="font-16">{city.current.visibility}</span>
        </li>
      </ul>
    </section>
  );
}
