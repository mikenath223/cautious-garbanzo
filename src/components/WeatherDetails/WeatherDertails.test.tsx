import { withRender } from "../../utils/testUtils";
import WeatherDetails from "./index";
import { defaultCity } from "../../utils/constants";

describe("Searchbar", function () {
  it("should match snapshot", () => {
    const { container } = withRender(
      <WeatherDetails
        city={defaultCity}
        isFavorites={false}
        onAddFavorites={() => jest.fn()}
        onRemoveFavorites={() => jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
