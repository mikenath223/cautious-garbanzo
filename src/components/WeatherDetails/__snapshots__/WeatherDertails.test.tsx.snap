// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar should match snapshot 1`] = `
<div>
  <section>
    <div
      class="weather-details"
    >
      <img
        alt=""
        class="icon-img"
        src="https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
      />
      <h1
        class="location-name"
      >
        New York
      </h1>
      <button
        class="submit-btn favorites-btn"
        type="submit"
      >
        <span>
          Add to favorites
        </span>
        <img
          alt=""
          class="icon-img"
          src="/assets/icons/unliked.png"
        />
      </button>
    </div>
    <div
      class="country-content flex-center"
    >
      <img
        alt=""
        class="icon-img"
        src="/assets/icons/country.png"
      />
      <span>
        United States of America
      </span>
    </div>
    <div
      class="flex-center"
    >
      <img
        alt=""
        class="icon-img"
        src="/assets/icons/state.png"
      />
      <span>
        New York
      </span>
    </div>
    <ul
      class="details-content"
    >
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Observation time
        </span>
        <span
          class="font-16"
        >
          12:14 PM
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Temperature
        </span>
        <span
          class="font-16"
        >
          13
          <sup>
            o
          </sup>
          C
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Weather description
        </span>
        <span
          class="font-16"
        >
          Sunny
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Wind speed
        </span>
        <span
          class="font-16"
        >
          0
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Wind degree
        </span>
        <span
          class="font-16"
        >
          349
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Wind direction
        </span>
        <span
          class="font-16"
        >
          N
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Pressure
        </span>
        <span
          class="font-16"
        >
          1010
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Precipitation
        </span>
        <span
          class="font-16"
        >
          0
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Humidity
        </span>
        <span
          class="font-16"
        >
          90
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Cloud cover
        </span>
        <span
          class="font-16"
        >
          0
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Feels like
        </span>
        <span
          class="font-16"
        >
          13
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          UV index
        </span>
        <span
          class="font-16"
        >
          4
        </span>
      </li>
      <li
        class="details-item card box-shadow"
      >
        <span
          class="font-20"
        >
          Visibility
        </span>
        <span
          class="font-16"
        >
          16
        </span>
      </li>
    </ul>
  </section>
</div>
`;
